### PR TITLE
core: support RISC-V syscall

### PR DESCRIPTION
**Pull request by liushiwei007**
Add 64-bit RISC-V system call function of connection.

Signed-off-by: liushiwei <liushiwei@eswincomputing.com>

### CODE REVIEW

**Comment by liushiwei007** [comment]
@jenswi-linaro @jforissier This submission does not appear to trigger a CI check

**Comment by jforissier** [comment]
@liushiwei007 indeed, that is unexpected. Neither GitHub action nor IBART did trigger so it may be a transient issue on GitHub's side. I suggest you try to force push again (do `git commit --amend` first to change the commit hash).

**Comment by jenswi-linaro** [comment]
This looks like a nice adaption of the arm counterpart. But it would be nice with a bit more surrounding code, for instance `user_ta_handle_svc()`. You don't need to populate the entire syscall table, but `syscall_sys_return()` and `syscall_panic()` seems given.

@maroueneboubakri, feel free to comment on these patches too.

**Comment by liushiwei007** [comment]
> But it would be nice with a bit more surrounding code, for instance `user_ta_handle_svc()`. You don't need to populate the entire syscall table, but `syscall_sys_return()` and `syscall_panic()` seems given.

Ok, I'll put in the surrounding code later

**Comment by liushiwei007** [comment]
@maroueneboubakri @jforissier  @jenswi-linaro Can you take a look at this submit

File: core/arch/riscv/tee/arch_ecall_rv64.S [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+/*
+ * uint32_t tee_ecall_do_call(struct thread_ctx_regs *regs, syscall_t func);
+ *
+ * Called from user_ta_handle_svc()
+ */
+FUNC tee_ecall_do_call , :
+UNWIND(	.cantunwind)
+	mv	t0, a0
+	mv	t1, a1
+	addi	sp, sp, -8
```
Comment (gagachang) [inline core/arch/riscv/tee/arch_ecall_rv64.S:19]: Should sp be 16-byte aligned?
(See latest RISC-V unprivileged ISA specification section 18.5 - Integer Computational Instructions)

File: core/arch/riscv/tee/arch_ecall_rv64.S [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	ld	a4, THREAD_REGS_A4(t0)
+	ld	a5, THREAD_REGS_A5(t0)
+	ld	a6, THREAD_REGS_A6(t0)
+	ld	a7, THREAD_REGS_A7(t0)
+	ld	a8, THREAD_REGS_A8(t0)
+
+	jalr	t1
+
+	ld ra, 0(sp)
+	addi sp, sp, 8
```
Comment (gagachang) [inline core/arch/riscv/tee/arch_ecall_rv64.S:35]: ditto

File: core/arch/riscv/tee/arch_ecall_rv64.S [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+ */
+
+/*
+ * User space sees this function as:
+ * void syscall_sys_return(uint32_t ret) __noreturn;
+ */
+FUNC syscall_sys_return , :
+	mv	a1, zero	/* panic = false */
+	mv	a2, zero	/* panic_code = 0 */
+	mv	a3, t0	/* pointer to struct thread_ctx_regs */
```
Comment (etienne-lms) [inline core/arch/riscv/tee/arch_ecall_rv64.S:52]: align inline comments indentation?

File: core/arch/riscv/tee/arch_ecall_rv64.S [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	j	tee_ecall_sys_return_helper
+END_FUNC syscall_sys_return
+
+/*
+ * User space sees this function as:
+ * void syscall_panic(uint32_t code) __noreturn;
+ */
+FUNC syscall_panic , :
+	addi a1, zero,1	/* panic = true */
```
Comment (etienne-lms) [inline core/arch/riscv/tee/arch_ecall_rv64.S:61]: tab after `addi`

**Review by etienne-lms — COMMENTED** [review]
some cosmetic concerns

**Comment by maroueneboubakri** [comment]
Hi @jenswi-linaro, @jforissier, @liushiwei007,

As there is much duplication on this code, basically `svc` has been replaced with `ecall`, I suggest to rename arch_svc.h|c to arch_syscall.h|c and move them to core/include/tee, core/tee and replace `thread_svc_regs` with `thread_syscall_regs` maybe ? and same applies for all SVC functions ?
Is this feasible ?

Regards
Maro

**Comment by jenswi-linaro** [comment]
@maroueneboubakri, that makes sense, I'll propose something.

**Comment by jenswi-linaro** [comment]
Please create a new pull request for "libutee: RISC-V syscall:fix bug".

**Comment by maroueneboubakri** [comment]
Stills not OK, remove the parentheses,  this way you are accessing memory at address \scn.

**Comment by liushiwei007** [comment]
> Stills not OK, remove the parentheses, this way you are accessing memory at address \scn.

ok

**Comment by liushiwei007** [comment]
> Please create a new pull request for "libutee: RISC-V syscall:fix bug".

ok

**Comment by liushiwei007** [comment]
@jenswi-linaro  @maroueneboubakri  hi,  is this submission OK？

**Comment by gagachang** [comment]
Dear @maroueneboubakri 
Please help review this minor fix then we can focus on other patches.

**Comment by liushiwei007** [comment]
@gagachang Is there anything else I should do for this submission?

**Comment by gagachang** [comment]
> @gagachang Is there anything else I should do for this submission?

No, I think this submission is independent from patches submitted by @maroueneboubakri .

**Comment by liushiwei007** [comment]
@jenswi-linaro, @jforissier  Can this submission be merged?

**Comment by jforissier** [comment]
Please use the `riscv` prefix in the commit subject. For example: `libutee: riscv: use li not mv in UTEE_SYSCALL`

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by maroueneboubakri** [comment]
> Please use the `riscv` prefix in the commit subject. For example: `libutee: riscv: use li not mv in UTEE_SYSCALL`
> 
> `Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

Also provide `num_args` in `t1`

**Comment by maroueneboubakri** [comment]
More comments:
- Provide `num_args`in `t1`
- This works also on RV32. Maybe rename `utee_syscalls_rv64.S` to `utee_syscalls_rv.S` and edit commit description.

`Reviewed-by: Marouene Boubakri <marouene.boubakri@nxp.com>`

**Comment by jenswi-linaro** [comment]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Pull request merged**